File: parse/bench/bench.ts
// Parser combinator benchmark in JavaScript: a JSON grammar on Parsimmon (see README.md).
import P from "parsimmon";

const ws = P.regexp(/[ \t\n\r]*/);
const tok = <T>(p: P.Parser<T>) => p.skip(ws);
const esc: Record<string, string> = { '"': '"', "\\": "\\", "/": "/", b: "\b", f: "\f", n: "\n", r: "\r", t: "\t" };
const str = P.regexp(/"((?:\\.|[^"\\\u0000-\u001f])*)"/, 1).map((s) =>
  s.replace(/\\(u[0-9a-fA-F]{4}|.)/g, (_, e) => (e.length === 5 ? String.fromCharCode(parseInt(e.slice(1), 16)) : esc[e])),
);

const J: P.Language = P.createLanguage({
  value: (r) => P.alt(r.object, r.array, r.string, r.number, r.null, r.true, r.false),
  lbrace: () => tok(P.string("{")),
  rbrace: () => tok(P.string("}")),
  lbracket: () => tok(P.string("[")),
  rbracket: () => tok(P.string("]")),
  comma: () => tok(P.string(",")),
  colon: () => tok(P.string(":")),
  null: () => tok(P.string("null")).result(null),
  true: () => tok(P.string("true")).result(true),
  false: () => tok(P.string("false")).result(false),
  string: () => tok(str),
  number: () => tok(P.regexp(/-?(0|[1-9][0-9]*)(\.[0-9]+)?([eE][+-]?[0-9]+)?/)).map(Number),
  array: (r) => r.lbracket.then(r.value.sepBy(r.comma)).skip(r.rbracket),
  pair: (r) => P.seq(r.string.skip(r.colon), r.value),
  object: (r) =>
    r.lbrace
      .then(r.pair.sepBy(r.comma))
      .skip(r.rbrace)
      .map((kv: [string, unknown][]) => Object.fromEntries(kv)),
});

function checksum(s: string): number {
  let h = 0;
  let n = 0;
  for (const c of s) {
    h = (Math.imul(h, 31) + c.codePointAt(0)!) >>> 0;
    n++;
  }
  return (h + n) >>> 0;
}

const doc = await Bun.file("out/doc.json").text();
const t0 = performance.now();
const v = ws.then(J.value).tryParse(doc);
const ms = performance.now() - t0;
// The document's keys are already sorted, and objects keep insertion order.
console.log(`parse\t${ms.toFixed(3)}\t${checksum(JSON.stringify(v))}`);

File: parse/bench/bench.py
"""Parser combinator benchmark in Python: a JSON grammar on pyparsing (see README.md)."""
import json
import time

import pyparsing as pp

LBRACK, RBRACK, LBRACE, RBRACE, COLON, COMMA = map(pp.Suppress, "[]{}:,")
NUMBER = pp.Regex(r"-?(0|[1-9][0-9]*)(\.[0-9]+)?([eE][+-]?[0-9]+)?").set_parse_action(
    lambda t: float(t[0]) if any(c in t[0] for c in ".eE") else int(t[0])
)
STRING = pp.QuotedString('"', esc_char="\\", convert_whitespace_escapes=True)
value = pp.Forward()
array = pp.Group(LBRACK + pp.Optional(pp.DelimitedList(value)) + RBRACK).set_parse_action(lambda t: [t[0].as_list()])
member = pp.Group(STRING + COLON + value)
obj = pp.Group(LBRACE + pp.Optional(pp.DelimitedList(member)) + RBRACE).set_parse_action(
    lambda t: [{k: v for k, v in t[0]}]
)
value <<= (
    pp.Keyword("null").set_parse_action(pp.replace_with(None))
    | pp.Keyword("true").set_parse_action(pp.replace_with(True))
    | pp.Keyword("false").set_parse_action(pp.replace_with(False))
    | NUMBER
    | STRING
    | array
    | obj
)


def checksum(s):
    h = 0
    for c in s:
        h = (h * 31 + ord(c)) & 0xFFFFFFFF
    return (h + len(s)) & 0xFFFFFFFF


doc = open("out/doc.json").read()
t0 = time.perf_counter()
v = value.parse_string(doc, parse_all=True)[0]
ms = (time.perf_counter() - t0) * 1e3
out = json.dumps(v, separators=(",", ":"), ensure_ascii=False, sort_keys=True)
print(f"parse\t{ms:.3f}\t{checksum(out)}")
